### PR TITLE
[MERGE] project sharing follow up

### PR DESCRIPTION
**Pull request by xavierbol**
Follow up of tasks 2379518 and 2633229 (#73341 and #77156).

Purpose
=======

The main goal is to allow the portal user to click on some stat button shown in the task form view in the project sharing feature. Those stat button will redirect to a portal view if the portal user has access to the document(s).

Then, some refactoring is made to easily use the existing code for one route to another one and review the code for the project sharing tour.

Finally, the `/my/project` and `/my/task` routes have been renamed to follow the same *naming* in the another routes. That is, those routes will be named `/my/projects` and `/my/tasks`.

## Details of the implementation

- Allow the portal user to click on the sale order stat button in task form view. This stat button will be clickable if the user has access to the sale order of the current task shown in the form view, otherwise the stat button will be invisible. If a portal user clicks on this button, then he will be redirect to `/my/orders/<id of the sale order>`.
- Allow to show the parent task from another project than the one shown in project sharing. If the portal user has access to the other project than there is no reason to not allow him to see the parent task.
- Rename the `/my/project` route into `/my/projects`.
- Rename the `/my/task` route into `/my/tasks`.
- Create hook methods to avoid duplicating code when we want to use this code to another route. It will be the case for a new route used in the project sharing feature.
- Create hook method in controller of account module to easily reuse the existing code for another route for project sharing feature.
- Refactor tour js for project sharing feature to remove the duplicating code.
- Revert the disable click on avatar. This changes was initially done for project sharing feature. However, this changes is made `web` module instead of the `project`. Since this changes is no longer used in project sharing feature and no used in backend code, the changes is reverted.
- Move the `invoice_count` field from `sale_project` module to `industry_fsm_sale` one because it seems this field is only used in that module and not anywhere before installing that module.

task-2648955

--
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Comment by IT-Ideas** [comment]
@robodoo merge

File: addons/sale_project/models/project.py [hunk 1]
```diff
@@ -212,6 +225,18 @@ def action_view_so(self):
...
 
+    def action_project_sharing_view_so(self):
+        self.ensure_one()
+        if self.user_has_groups('base.group_portal'):
+            if not self.display_sale_order_button:
+                return {}
+            return {
+                "name": "Portal Sale Order",
+                "type": "ir.actions.act_url",
+                "url": "/my/orders/%s" % self.sale_order_id.id,
```
Comment (Feyensv) [inline addons/sale_project/models/project.py:236]:
```suggestion
                "url": self.sale_order_id.access_url,
```
?

**Review by Feyensv — COMMENTED** [review]
lgtm for sale part :ok_hand: 
But I'm a little bit concerned by the change of url for project/task portal (from `/my/task` to `/my/tasks` & `/my/project` to `/my/projects`).  This is the kind of change that will break all links received by customers before a migration :thinking:.
If this change is really necessary, shouldn't you keep both urls and drop the old one in one year or sthg like that ?

**Comment by AntoineVDV** [comment]
> lgtm for sale part ok_hand But I'm a little bit concerned by the change of url for project/task portal (from `/my/task` to `/my/tasks` & `/my/project` to `/my/projects`). This is the kind of change that will break all links received by customers before a migration thinking. If this change is really necessary, shouldn't you keep both urls and drop the old one in one year or sthg like that ?

There is no such thing as "in one year" when you consider that a database can migrate from Odoo 15.0 to Odoo 17.0 overnight. The link would still be broken even if you supported the old route from 2022 to 2023.

**Comment by xavierbol** [comment]
@AntoineVDV @Feyensv 
Thanks for the review and your opinion about the change `/my/task/<id of the task>` => `/my/tasks/<id of the task>` and `/my/project/<id of the project>` => `/my/projects<id of the project>`. Obviously, this is not really necessary, the idea of it is to uniform the portal routes to have the same logic than the other routes. 
Currently, when the user sees the list of projects, the route is `/my/projects` and when he selects a project, the route becomes `/my/project/<id of the project>`, I find a bit pity we don't conserve the initial route, that is `/my/projects` and just add the id of the selected project in the route to shown the portal form view.
Like your routes in Sales, it is `/my/orders` to list all the available SOs for current user and when the user clicks to a order, the route will be `/my/orders/<id of the order selected>` and not `/my/order/<id of the order selected>`. 🙂 

Indeed, II agree with you, it will break all links to the customers, so it is not a good idea, I will revert my commits about that change. 🙂

cc @IT-Ideas

**Comment by AntoineVDV** [comment]
> I will revert my commits about that change.

Alternatively, you could support both routes at once in the controller but use `/projects` and `/tasks` everywhere else. See for example: https://github.com/odoo/odoo/blob/master/addons/payment_ogone/controllers/main.py#L18-L26

File: addons/project/controllers/portal.py [hunk 2]
```diff
@@ -63,7 +63,7 @@ def _project_get_page_view_values(self, project, access_token, page=1, date_begi
         # task count
         task_count = Task.search_count(domain)
         # pager
-        url = "/my/project/%s" % project.id
+        url = "/my/projects/%s" % project.id
```
Comment (IT-Ideas) [inline addons/project/controllers/portal.py:36]: I think we should keep the current routes and call the new ones from there. We could then remove the old ones when it seems reasonable that no mail message will remain used with the old refs.

File: addons/project/views/project_sharing_views.xml [hunk 3]
```diff
@@ -103,7 +103,7 @@
                                     </span>
                                     <field name="kanban_state" widget="state_selection"/>
                                     <!-- TODO: [XBO] remove me in master -->
-                                    <field name="user_ids" string="Assignees" widget="many2many_avatar_user" options="{'no_open_chat': True}" invisible="1"/>
+                                    <field name="user_ids" string="Assignees" widget="many2many_avatar_user" invisible="1"/>
```
Comment (IT-Ideas) [inline addons/project/views/project_sharing_views.xml:106]: Why don't you remove the field?

File: addons/sale_project/models/project.py [hunk 4]
```diff
@@ -177,6 +179,18 @@ def _compute_sale_line(self):
...
 
+    @api.depends('sale_order_id')
+    def _compute_display_sale_order_button(self):
+        if not self.sale_order_id:
+            self.display_sale_order_button = False
+            return
+        try:
+            sale_orders = self.env['sale.order'].search([('id', 'in', self.sale_order_id.ids)])
+            for task in self:
+                task.display_sale_order_button = task.sale_order_id and task.sale_order_id in sale_orders
```
Comment (IT-Ideas) [inline addons/sale_project/models/project.py:190]: Do we really need to check that `task.sale_order_id` is not Falsy prior to check if in `sale_orders`?

**Review by IT-Ideas — COMMENTED** [review]
Thanks for your work :)
Just a bit concerned with the routes and as @odony proposed me this morning as I chatted with him about that, we could just keep the old ones and call the new ones. Once we are confident enough that no more mail message containing those old routes will be used, we will drop them.

**Comment by IT-Ideas** [comment]
> > I will revert my commits about that change.
> 
> Alternatively, you could support both routes at once in the controller but only use `/projects` and `/tasks` everywhere else. See for example: https://github.com/odoo/odoo/blob/master/addons/payment_ogone/controllers/main.py#L18-L26

Oups I started the review early this morning but was disturbed and could not complete it directly. I agree with you and @Feyensv about the concern with the routes 🙂 I had the same. Thanks for your review and @Feyensv 's one too!

File: addons/project/controllers/portal.py [hunk 5]
```diff
@@ -146,20 +90,22 @@ def portal_my_projects(self, page=1, date_begin=None, date_end=None, sortby=None
         })
         return request.render("project.portal_my_projects", values)
 
-    @http.route(['/my/project/<int:project_id>'], type='http', auth="public", website=True)
```
[the remark this answers is not in the document]
Comment (AntoineVDV) [reply addons/project/controllers/portal.py:149]: https://github.com/odoo/odoo/pull/82379#pullrequestreview-864600291

File: addons/project/views/project_sharing_views.xml (hunk 3 above)
Comment (xavierbol) [reply addons/project/views/project_sharing_views.xml:106]: It is removed in #77024. 🙂

File: addons/project/controllers/portal.py [hunk 6]
```diff
@@ -146,20 +90,35 @@ def portal_my_projects(self, page=1, date_begin=None, date_end=None, sortby=None
...
+                 '/my/project/<int:project_id>/page/<int:page>',
+                 '/my/project/<int:project_id>/task/<int:task_id>',
+                 '/my/project/<int:project_id>/project_sharing',
+                 '/my/task',
+                 '/my/task/page/<int:page>'], type='http', auth="public")
+    def portal_project_routes_outdated(self, **kwargs):
+        """Redirect the outdated routes to the new routes."""
+        path = request.httprequest.path
+        if path.startswith('/my/project/'):
+            return request.redirect(path.replace('/my/project/', '/my/projects/'))
```
Comment (Feyensv) [inline addons/project/controllers/portal.py:103]: Might be worth two routes no ? :D

**Review by Feyensv — APPROVED** [review]

File: addons/project/controllers/portal.py (hunk 6 above)
Comment (xavierbol) [reply addons/project/controllers/portal.py:103]: 2 methods? One for project and one for task? 🙂 
Why not, it is the same for me. ^^

Comment (Feyensv) [reply addons/project/controllers/portal.py:103]: Yep, but do as you like :D

Comment (xavierbol) [reply addons/project/controllers/portal.py:103]: Ok, thanks for your reviews! 😉

**Review by IT-Ideas — COMMENTED** [review]
Thanks for your work, just two small things.

**Review by IT-Ideas — APPROVED** [review]

**Comment by IT-Ideas** [comment]
@robodoo r+

**Pull request merged**